[Sean Spencer Manik] iP

seanmanik opened the pull request:
# DukePro
**_DukePro_** serves as an easy-to-use and powerful task list that supports the addition of:
- Events
- Deadlines
- Todos

Here are the _benefits_ that you will be able to derive from **DukePro**:
1. Never miss a deadline ever again.
2. Mark tasks whenever you are done, which will motivate you to want to finish off everything in your task list!
3. Keep track of dates relating to your deadlines and events!

Here's a piece of code that says hello to you!
```java
public class Main {
    public static void main(String [] args) {
        System.out.println("Hello!");
    }
}
```
Upcoming features:
- [x] Save task list :open_file_folder:
- [ ] Help you do your work :book:
- [ ] Tell jokes :grinning:

My favourite quote is:
> "Hard work beats talent when talent doesn't work hard." [-Tim Notke](https://www.lifehack.org/articles/productivity/hard-work-beats-talent-when-talent-doesnt-work-hard.html)

Commit:
Add Gradle support

seanmanik committed:
Level-2 completed.

seanmanik committed:
Revert "Level-2 completed."

This reverts commit 8a4241894e6f5e058bcd1150766277953adada9d.

seanmanik committed:
Added Level-1 code.

seanmanik committed:
Bot is now able to record responses in a list, and return them when requested. Level-2 added.

seanmanik committed:
Mark and unmark functions added. Level-3 done.

seanmanik committed:
Todo, Deadline, and Event has been added. Level-4 completed.

seanmanik committed:
Saved test files.

seanmanik committed:
Error handling added. Level-5 completed.

seanmanik committed:
Deletion function added. Level-6 completed.

seanmanik committed:
Error handling updated to be handled with message shown to user.

seanmanik committed:
Change greet function.

seanmanik committed:
Add saving and file read functions. Level-7 completed.

seanmanik committed:
Add date formatter. Level-8 completed.

seanmanik committed:
Merge branch 'branch-Level-8'

seanmanik committed:
Create SaveFile class

seanmanik committed:
Add Ui, Storage, TaskList classes. A-MoreOOP Completed.

seanmanik committed:
Add duke package. A-Packages done.

seanmanik committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

seanmanik committed:
Add JUnit tests for Deadline and TaskList.

seanmanik committed:
Add javadocs to all classes and methods.

seanmanik committed:
Change Coding Standards

seanmanik committed:
Add find function. Level-9 completed.

seanmanik committed:
Merge branch 'branch-A-CodingStandard'

seanmanik committed:
Merge branch 'branch-Level-9'

seanmanik committed:
Add Parser Class

seanmanik committed:
Merge branch 'branch-A-JavaDoc'

seanmanik committed:
Add more OOP, Command and Parser classes implemented.

seanmanik committed:
Add updated jar file

seanmanik committed:
Add checkstyle

seanmanik committed:
Add JavaFX Template

seanmanik committed:
GUI Implemented Level-10 completed!

L1uY1jun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,79 @@
+            if (info.length <= 1) {
+                throw new DukeException("OOPS!! The description of a todo cannot be empty.");
+            }
+            return new AddTaskCommand(tasks, new ToDo(info[1]), ui);
+        } else if (input.startsWith("deadline")) {
+            String[] info = input.split("deadline ");
+            if (info.length <= 1) {
+                throw new DukeException("OOPS!! The description of a deadline cannot be empty.");
+            }
+            String[] item = info[1].split(" /by ");
+            if (item.length <= 1) {
+                throw new DukeException("OOPS!! A deadline has to be set!");
+            }
+            return new AddTaskCommand(tasks, new Deadline(item[0], item[1]), ui);
+        } else if (input.startsWith("event")) {
+            String[] info = input.split("event ");
+            if (info.length <= 1) {
+                throw new DukeException("OOPS!! The description of a event cannot be empty.");
+            }
+            String[] item = info[1].split(" /at ");
+            if (item.length <= 1) {
+                throw new DukeException("OOPS!! The timing of the event has to be set!");
+            }
+            return new AddTaskCommand(tasks, new Event(item[0], item[1]), ui);
+        } else if (input.startsWith("delete")) {
+            String[] info = input.split("delete ");
+            if (info.length <= 1) {
+                throw new DukeException("OOPS!! The item to be deleted has to be specified.");
+            }
+            int target = Integer.valueOf(info[1]) - 1;
+            return new DeleteCommand(tasks, target, ui);
+        } else if (input.startsWith("find")) {
+            String[] info = input.split("find ");
+            if (info.length <= 1) {
+                throw new DukeException("Please specify what you would like to find.");
+            }
+            return new FindCommand(tasks, info[1], ui);
+        } else {
+            return new UnknownCommand(ui);
+        }
```
Recommended to check out String.strip() method under the String class [here](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#strip()) so you would not run into problems with whitespaces.

L1uY1jun commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+            Scanner reader = new Scanner(this.file);
+            while (reader.hasNextLine()) {
+                String info = reader.nextLine();
+                if (info.startsWith("  [T]")) {
+                    String[] infoData = info.split("] ");
+                    ToDo toAdd = new ToDo(infoData[infoData.length - 1]);
+                    if (isMarked(infoData[infoData.length - 2])) {
+                        toAdd.mark();
+                    }
+                    ls.add(toAdd);
+                } else if (info.startsWith("  [D]")) {
+                    String[] infoData = info.split("] ");
+                    String[] descBy = infoData[infoData.length - 1].split("by: ");
+                    Deadline toAdd = new Deadline(descBy[0].substring(0, descBy[0].length() - 2),
+                            descBy[1].substring(0, descBy[1].length() - 1));
+                    if (isMarked(infoData[infoData.length - 2])) {
+                        toAdd.mark();
+                    }
+                    ls.add(toAdd);
+                } else if (info.startsWith("  [E]")) {
+                    String[] infoData = info.split("] ");
+                    String[] descAt = infoData[infoData.length - 1].split("at: ");
+                    Event toAdd = new Event(descAt[0].substring(0, descAt[0].length() - 2),
+                            descAt[1].substring(0, descAt[1].length() - 1));
+                    if (isMarked(infoData[infoData.length - 2])) {
+                        toAdd.mark();
```
Similar recommendation here to check out String.strip() method for easier debugging rather than substring method.

L1uY1jun commented on src/main/java/dukepro/DukePro.java:
```diff
@@ -0,0 +1,158 @@
+        return "You: " + input;
+    }
+
+    public void startupDuke() {
```
Naming issue for method, should be `startUpDuke` in camelCase

L1uY1jun commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,26 @@
+public class DeadlineTest {
+
+    @Test
+    public void incorrectDateTest() {
```
verb for method, perhaps `testIncorrectDate` is better

L1uY1jun commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,27 @@
+public class TaskListTest {
+
+    @Test
+    public void addTest() {
```
Similar code quality problem here, verb for command, suggestion: `testAddTask`

L1uY1jun commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @Test
+    public void deleteTest() {
```
Similar code quality problem here, verb for command, suggestion: `testDeleteTask`

L1uY1jun reviewed (commented):
Very few code quality issues spotted, overall consistent naming. Good effort!
I have included some other suggestions you could look through 👍.

seanmanik committed:
Add new JAR file

JAR file now runs Duke that has a JavaFX-created GUI instead
of the previous command line interface.

seanmanik committed:
Add assertions.

Assertions were added to check if arrays obtained after using the
split() method were of the right length.

seanmanik committed:
Improve Code Quality and readability.

Javadocs were added for all necessary methods and classes.

seanmanik committed:
Merge pull request #1 from seanmanik/branch-A-Assertions

Add assertions.

seanmanik committed:
Merge pull request #2 from seanmanik/branch-A-CodeQuality

Improve Code Quality and readability.

seanmanik committed:
Add extra javadocs.

seanmanik committed:
Add update capabilities.

Users will now be able to use the update command to
update their tasks.

ToDo tasks can have their description updated, Deadline
tasks can have their description and date updated, and
Event tasks can have their description and time updated.

seanmanik committed:
Add .yml file for CI purposes.

seanmanik committed:
Add Ui.png

seanmanik committed:
Add User Guide

User guide was added, with features and explanations of how to run DukePro.

seanmanik committed:
Add support for jar files.

Now, jar file will automatically create a folder for storing
Duke.txt data if the folders and files are not present.

seanmanik committed:
Merge branch 'master' of https://github.com/seanmanik/ip

seanmanik committed:
Switch to JavaFX 11

seanmanik committed:
Update User Guide
